asdf/extension.py

```diff
@@ -107,24 +107,88 @@ def url_mapping(self):
         return []
 
 
+class ExtensionProxy(AsdfExtension):
+    """
+    Proxy that wraps an `AsdfExtension`, provides default
+    implementations of optional methods, and carries additional
+    information on the package that provided the extension.
+    """
+    @classmethod
+    def maybe_wrap(self, delegate):
+        if isinstance(delegate, ExtensionProxy):
+            return delegate
+        else:
+            return ExtensionProxy(delegate)
+
+    def __init__(self, delegate, package_name=None, package_version=None, legacy=False):
+        if not isinstance(delegate, AsdfExtension):
+            raise TypeError("Extension must implement the AsdfExtension interface")
+
+        self._delegate = delegate
+        self._package_name = package_name
+        self._package_version = package_version
+        self._legacy = legacy
+
+        self._class_name = get_class_name(delegate)
+
+    @property
+    def types(self):
+        return getattr(self._delegate, "types", [])
+
+    @property
+    def tag_mapping(self):
+        return getattr(self._delegate, "tag_mapping", [])
+
+    @property
+    def url_mapping(self):
+        return getattr(self._delegate, "url_mapping", [])
+
+    @property
+    def delegate(self):
+        return self._delegate
+
+    @property
+    def class_name(self):
+        return self._class_name
+
+    @property
+    def package_name(self):
+        return self._package_name
+
+    @property
+    def package_version(self):
+        return self._package_version
+
+    @property
+    def legacy(self):
+        return self._legacy
+
+    def __repr__(self):
+        return "<ExtensionProxy class: {} package: {}=={} legacy: {}>".format(
+            self.class_name,
+            self.package_name,
+            self.package_version,
+            self.legacy,
+        )
+
+
 class AsdfExtensionList:
     """
     Manage a set of extensions that are in effect.
     """
     def __init__(self, extensions):
+        self._extensions = [ExtensionProxy.maybe_wrap(e) for e in extensions]
         tag_mapping = []
         url_mapping = []
         validators = {}
         self._type_index = AsdfTypeIndex()
-        for extension in extensions:
+        for extension in self._extensions:
             if not isinstance(extension, AsdfExtension):
                 raise TypeError(
                     "Extension must implement asdf.types.AsdfExtension "
                     "interface")
             tag_mapping.extend(extension.tag_mapping)
-            # New-style extensions will not include a url_mapping attribute.
-            if hasattr(extension, "url_mapping"):
-                url_mapping.extend(extension.url_mapping)
+            url_mapping.extend(extension.url_mapping)
```

Comment:
This is the benefit of wrapping the extensions in that `ExtensionProxy` object, we don't have to make these `hasattr` checks all over the place.  We're going to have more and more optional and deprecated properties on these things and I think this will save some sanity.